<a href="https://colab.research.google.com/github/GuillermoAlmandoz/IEBS-Master-Data-Science/blob/main/Proyecto_M1_Copa_del_mundo_Git2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://raw.githubusercontent.com/GuillermoAlmandoz/IEBS-Master-Data-Science/main/notebook_banner.png" width="1024" style="float:center">
<div style="margin-top:10px; display:flex; align-items:center;">
    <div style="float:left; margin-right:10px;vertical-align:middle">
        <a href="https://www.linkedin.com/in/Guillermo-Almandoz-Rios/" targer="_blank">
            <img src="https://raw.githubusercontent.com/GuillermoAlmandoz/IEBS-Master-Data-Science/main/lkin-logo.png"/>
        </a>
    </div>
    <div> 
        Guillermo Almándoz (23/01/2023)
    </div>
 </div>

# Dataset

En esta libreta vamos a trabajar con un conjunto de datos obtenido de [Kaggle](https://www.kaggle.com/abecklas/fifa-world-cup?select=WorldCupMatches.csv) que contiene los datos de la FIFA de la copa del mundo. Recomendamos acceder a este enlace para descargar los ficheros a trabajar, aunque también disponéis de los archivos ya descargados en la sección de recursos. Utilizaremos estos tres ficheros:

- WorldCupMatches.csv: Datos de los partidos disputados
- WorldCupPlayers.csv: Datos de los jugadores que han jugado en cada partido
- WorldCups.csv: Datos de las copas disputadas y los resultados

# Introducción

Mediante esta libreta vamos a explorar los datos almacenados en estos tres ficheros y a responder preguntas haciendo operaciones y transformaciones con pandas. Además de completar las celdas de código para responder a los ejercicios, cuando así se indique se deberá añadir una celda de tipo _markdown_ explicando con texto enriquecido la información o las conclusiones extraídas.

In [ ]:
import pandas as pd

In [ ]:
# Recolectamos los datos desde el dataset de la plataforma Kaggle abecklas/fifa-world-cup
# Importamos fichero local previamente descargado el ficero
from google.colab import files
uploaded=files.upload()


Saving WorldCupMatches.csv to WorldCupMatches.csv
Saving WorldCupPlayers.csv to WorldCupPlayers.csv
Saving WorldCups.csv to WorldCups.csv


In [ ]:
# Cargamos el conjunto de datos WorldCupMatches
df_partidos = pd.read_csv("WorldCupMatches.csv")

**NOTA:** en una actividad anterior completamos los ejercicios del 1 al 3, copiar y pegar dicho trabajo en esta libreta con el fin de tener un report completo que analize los datos del problema. Es decir, **no es necesario volver a realizar estos 3 primeros ejercicios**.

## 1. Exploración

Usar las funciones `head`, `info` y `describe` para explorar el dataframe `df_partidos`. **Crear una celda de tipo markdown** y comentar brevemente la información almacenada en cada dataframe (columnas y tipo de datos).

_No es necesario hacer una documentación exhaustiva, tan solo comentar aspectos principales como qué tipo de información almacena el DataFrame, cuántas filas y columnas tiene, y comentar las columnas que se consideren más relevantes._

In [ ]:
# Copiar solución de la actividad semanal correspondiente
# exploración del dataset
df_partidos

Year              Datetime    Stage         Stadium         City  \
0     1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1     1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2     1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3     1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4     1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    
...      ...                   ...      ...             ...          ...   
4567     NaN                   NaN      NaN             NaN          NaN   
4568     NaN                   NaN      NaN             NaN          NaN   
4569     NaN                   NaN      NaN             NaN          NaN   
4570     NaN                   NaN      NaN             NaN          NaN   
4571     NaN                   NaN      NaN             NaN          NaN   

     Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0            France              4.0              1.0         Mexico   
1               USA              3.0              0.0        Belgium   
2        Yugoslavia              2.0              1.0         Brazil   
3           Romania              3.0              1.0           Peru   
4         Argentina              1.0              0.0         France   
...             ...              ...              ...            ...   
4567            NaN              NaN              NaN            NaN   
4568            NaN              NaN              NaN            NaN   
4569            NaN              NaN              NaN            NaN   
4570            NaN              NaN              NaN            NaN   
4571            NaN              NaN              NaN            NaN   

     Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                        4444.0                   3.0                   0.0   
1                       18346.0                   2.0                   0.0   
2                       24059.0                   2.0                   0.0   
3                        2549.0                   1.0                   0.0   
4                       23409.0                   0.0                   0.0   
...             ...         ...                   ...                   ...   
4567            NaN         NaN                   NaN                   NaN   
4568            NaN         NaN                   NaN                   NaN   
4569            NaN         NaN                   NaN                   NaN   
4570            NaN         NaN                   NaN                   NaN   
4571            NaN         NaN                   NaN                   NaN   

                     Referee               Assistant 1  \
0     LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1          MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2        TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3      WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4        REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   
...                      ...                       ...   
4567                     NaN                       NaN   
4568                     NaN                       NaN   
4569                     NaN                       NaN   
4570                     NaN                       NaN   
4571                     NaN                       NaN   

                     Assistant 2  RoundID  MatchID Home Team Initials  \
0            REGO Gilberto (BRA)    201.0   1096.0                FRA   
1          WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2            BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3       MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4     RADULESCU Constantin (ROU)    201.0   1085.0                ARG   
...                          ...      ...      ...                ...   
4567                         NaN      NaN      NaN                NaN   


In [ ]:
#exploracion de las primeras lineas del dataset
df_partidos.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   
3                     2549.0                   1.0                   0.0   
4                    23409.0                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
0         REGO Gilberto (BRA)    201.0   1096.0                FRA   
1       WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2         BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3    MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4  RADULESCU Constantin (ROU)    201.0   1085.0                ARG   

  Away Team Initials  
0                MEX  
1                BEL  
2                BRA  
3                PER  
4                FRA

In [ ]:
#consultamos datos estadísticos generales:número de elementos, media, desviación standard, valor mínimo, percentiles 25%,50%,75%, valor máximo
df_partidos.describe(include="all").transpose()

count unique                       top freq  \
Year                  852.0    NaN                       NaN  NaN   
Datetime                852    602      27 May 1934 - 16:30     8   
Stage                   852     23               Round of 16   72   
Stadium                 852    181            Estadio Azteca   19   
City                    852    151              Mexico City    23   
Home Team Name          852     78                    Brazil   82   
Home Team Goals       852.0    NaN                       NaN  NaN   
Away Team Goals       852.0    NaN                       NaN  NaN   
Away Team Name          852     83                    Mexico   38   
Win conditions          852     43                            787   
Attendance            850.0    NaN                       NaN  NaN   
Half-time Home Goals  852.0    NaN                       NaN  NaN   
Half-time Away Goals  852.0    NaN                       NaN  NaN   
Referee                 852    366     Ravshan IRMATOV (UZB)   10   
Assistant 1             852    387      ACHIK Redouane (MAR)    7   
Assistant 2             852    408  KOCHKAROV Bakhadyr (KGZ)   10   
RoundID               852.0    NaN                       NaN  NaN   
MatchID               852.0    NaN                       NaN  NaN   
Home Team Initials      852     77                       BRA   82   
Away Team Initials      852     82                       MEX   38   

                                 mean               std     min      25%  \
Year                      1985.089202         22.448825  1930.0   1970.0   
Datetime                          NaN               NaN     NaN      NaN   
Stage                             NaN               NaN     NaN      NaN   
Stadium                           NaN               NaN     NaN      NaN   
City                              NaN               NaN     NaN      NaN   
Home Team Name                    NaN               NaN     NaN      NaN   
Home Team Goals              1.811033          1.610255     0.0      1.0   
Away Team Goals                1.0223          1.087573     0.0      0.0   
Away Team Name                    NaN               NaN     NaN      NaN   
Win conditions                    NaN               NaN     NaN      NaN   
Attendance                    45164.8      23485.249247  2000.0  30000.0   
Half-time Home Goals          0.70892          0.937414     0.0      0.0   
Half-time Away Goals         0.428404          0.691252     0.0      0.0   
Referee                           NaN               NaN     NaN      NaN   
Assistant 1                       NaN               NaN     NaN      NaN   
Assistant 2                       NaN               NaN     NaN      NaN   
RoundID               10661772.591549    27296131.70287   201.0    262.0   
MatchID               61346867.552817  111057171.671912    25.0  1188.75   
Home Team Initials                NaN               NaN     NaN      NaN   
Away Team Initials                NaN               NaN     NaN      NaN   

                          50%          75%          max  
Year                   1990.0       2002.0       2014.0  
Datetime                  NaN          NaN          NaN  
Stage                     NaN          NaN          NaN  
Stadium                   NaN          NaN          NaN  
City                      NaN          NaN          NaN  
Home Team Name            NaN          NaN          NaN  
Home Team Goals           2.0          3.0         10.0  
Away Team Goals           1.0          2.0          7.0  
Away Team Name            NaN          NaN          NaN  
Win conditions            NaN          NaN          NaN  
Attendance            41579.5      61374.5     173850.0  
Half-time Home Goals      0.0          1.0          6.0  
Half-time Away Goals      0.0          1.0          5.0  
Referee                   NaN          NaN          NaN  
Assistant 1               NaN          NaN          NaN  
Assistant 2               NaN          NaN     


Los campos del Dataframe que disponemos son de **4572** filas y **20** columnas.
Corresponden a los partidos **852** filas, ya que el resto de valores son nulos. Tenemos **3720** filas con valores perdidos.

Disponemos de variables que se corresponden con la fecha de los partidos:

*   Year
*   Date time

Columnas relacionadas con la localidad dónde se jugó el partido:

*   Stage
*   Stadium
*   City

Y otras variables relacionadas con el partido en sí, como los equipos implicados (`Home Team Name` y `Away Team Name` ), sus iniciales (`Home Team Name` y `Away Team Name`) y el resultado del partido (`Home Team Goals` y `Away Team Goals`).


Los campos corresponden a las siguientes informaciones:

0 Year: Año en el que se celebró la copa del mundo, elcampo es numérico

1 Datetime: El día y la hora exacta que se jugó el partido, el campo es un objecto.

2 Stage: La fase de clasificación que se jugó el partido, el campo es objeto

3 Stadium: El estadio donde se jugó el partido, el campo es objeto

4 City: La ciudad donde se jugó el partido, el campo es objeto

5 Home Team Name: El equipo que jugaba en casa, el campo es objeto

6 Home Team Goal: El número de goles que marcó el equipo que jugaba en casa, el campo es numérico

7 Away Team Name: El equipo que jugaba fuera de casa, el campo es objeto

8 Away Team Goal: El número de goles que marcó el equipo que jugaba fuera de casa, el campo es numérico

9 Win conditions: Si hubo tiempo de descuento en el partido, el campo es objeto

10 Attendance: El público en las gradas, el campo es numérico

11 Half-time Home Goals: Los goles que se marcaron en la primera parte por el equipo de casa, el campo es numérico

12 Half-time Away Goals: El número de goles que marcó el equipo que jugaba fuera de casa, el campo es numérico

13 Referee: El nombre del árbitro, el campo es objeto

14 Assistant 1: El nombre del primer asistente, el campo es objeto

15 Assistant 2: El nombre del segundo asistente, el campo es objeto

16 RoundID: El identificador de la ronda de partido, el campo es numérico

17 MatchID: El identificador del partido, el campo es numérico

18 Home Team Initials: Las iniciales del equipo de casa, el campo es objeto

19 Away Team Initials: Las iniciales del equipo fuera de casa, el campo es objeto

## 2. Limpieza de datos

Una de las cosas que llama la atención es la alta presencia de valores perdidos en el DataFrame `df_partidos`. Cuando trabajamos con datos obtenidos del mundo real siempre nos toparemos con problemas relacionados con la medición, captura o almacenamiento de dicha información.

Localiza las filas con valores perdido. Analizar y **documentar en una nueva celda de tipo markdown** a qué se deben estos valores perdidos. Finalmente usar la función [dropna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) sobre el DataFrame `df_partidos` para eliminar los valores perdidos del DataFrame, y almacena el resultado en `df_partidos` de nuevo.

In [ ]:
# Copiar solución de la actividad semanal correspondiente
# Como vimos en el describe anterior, solo la columna Attendance tiene 2 valores perdidos más que el resto
# Usamos esta columna para comprobar cuando es NA y ver las filas correspondientes
df_partidos.loc[df_partidos.Attendance.isna()].head()

Year              Datetime        Stage            Stadium  \
823  2014.0  30 Jun 2014 - 17:00   Round of 16  Estadio Beira-Rio   
841  2014.0  30 Jun 2014 - 17:00   Round of 16  Estadio Beira-Rio   
852     NaN                   NaN          NaN                NaN   
853     NaN                   NaN          NaN                NaN   
854     NaN                   NaN          NaN                NaN   

              City Home Team Name  Home Team Goals  Away Team Goals  \
823  Porto Alegre         Germany              2.0              1.0   
841  Porto Alegre         Germany              2.0              1.0   
852            NaN            NaN              NaN              NaN   
853            NaN            NaN              NaN              NaN   
854            NaN            NaN              NaN              NaN   

    Away Team Name                 Win conditions  Attendance  \
823        Algeria  Germany win after extra time          NaN   
841        Algeria  Germany win after extra time          NaN   
852            NaN                            NaN         NaN   
853            NaN                            NaN         NaN   
854            NaN                            NaN         NaN   

     Half-time Home Goals  Half-time Away Goals             Referee  \
823                   0.0                   0.0  RICCI Sandro (BRA)   
841                   0.0                   0.0  RICCI Sandro (BRA)   
852                   NaN                   NaN                 NaN   
853                   NaN                   NaN                 NaN   
854                   NaN                   NaN                 NaN   

                   Assistant 1              Assistant 2   RoundID  \
823  DE CARVALHO Emerson (BRA)  VAN GASSE Marcelo (BRA)  255951.0   
841  DE CARVALHO Emerson (BRA)  VAN GASSE Marcelo (BRA)  255951.0   
852                        NaN                      NaN       NaN   
853                        NaN                      NaN       NaN   
854                        NaN                      NaN       NaN   

         MatchID Home Team Initials Away Team Initials  
823  300186460.0                GER                ALG  
841  300186460.0                GER                ALG  
852          NaN                NaN                NaN  
853          NaN                NaN                NaN  
854          NaN                NaN                NaN

In [ ]:
# Ejemplo de cómo ver las filas con valores perdidos en el código cuando alguna de las columnas tiene NAs
df_partidos.loc[df_partidos.isna().any(axis=1)].head()

Year              Datetime        Stage            Stadium  \
823  2014.0  30 Jun 2014 - 17:00   Round of 16  Estadio Beira-Rio   
841  2014.0  30 Jun 2014 - 17:00   Round of 16  Estadio Beira-Rio   
852     NaN                   NaN          NaN                NaN   
853     NaN                   NaN          NaN                NaN   
854     NaN                   NaN          NaN                NaN   

              City Home Team Name  Home Team Goals  Away Team Goals  \
823  Porto Alegre         Germany              2.0              1.0   
841  Porto Alegre         Germany              2.0              1.0   
852            NaN            NaN              NaN              NaN   
853            NaN            NaN              NaN              NaN   
854            NaN            NaN              NaN              NaN   

    Away Team Name                 Win conditions  Attendance  \
823        Algeria  Germany win after extra time          NaN   
841        Algeria  Germany win after extra time          NaN   
852            NaN                            NaN         NaN   
853            NaN                            NaN         NaN   
854            NaN                            NaN         NaN   

     Half-time Home Goals  Half-time Away Goals             Referee  \
823                   0.0                   0.0  RICCI Sandro (BRA)   
841                   0.0                   0.0  RICCI Sandro (BRA)   
852                   NaN                   NaN                 NaN   
853                   NaN                   NaN                 NaN   
854                   NaN                   NaN                 NaN   

                   Assistant 1              Assistant 2   RoundID  \
823  DE CARVALHO Emerson (BRA)  VAN GASSE Marcelo (BRA)  255951.0   
841  DE CARVALHO Emerson (BRA)  VAN GASSE Marcelo (BRA)  255951.0   
852                        NaN                      NaN       NaN   
853                        NaN                      NaN       NaN   
854                        NaN                      NaN       NaN   

         MatchID Home Team Initials Away Team Initials  
823  300186460.0                GER                ALG  
841  300186460.0                GER                ALG  
852          NaN                NaN                NaN  
853          NaN                NaN                NaN  
854          NaN                NaN                NaN

In [ ]:
df_partidos.loc[df_partidos.isna().any(axis=1)].shape

(3722, 20)

Vemos que las últimas **3720** filas contienen únicamente valores perdidos. Esto puede deberse a la forma de generar el fichero csv, bien por un error al generarlo de forma programática sujeto a un número prefijado de filas, o bien por un error humano, por ejemplo al usar un software de hojas de cálculo, seleccionando de forma incorrecta las filas a exportar.

In [ ]:
# Eliminamos las filas con algún valor perdido
df_partidos = df_partidos.dropna()

In [ ]:
#exploracion del tipo de variables y las columnas, así como los valores no-nulos y uso de memoria
df_partidos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 850 entries, 0 to 851
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  850 non-null    float64
 1   Datetime              850 non-null    object 
 2   Stage                 850 non-null    object 
 3   Stadium               850 non-null    object 
 4   City                  850 non-null    object 
 5   Home Team Name        850 non-null    object 
 6   Home Team Goals       850 non-null    float64
 7   Away Team Goals       850 non-null    float64
 8   Away Team Name        850 non-null    object 
 9   Win conditions        850 non-null    object 
 10  Attendance            850 non-null    float64
 11  Half-time Home Goals  850 non-null    float64
 12  Half-time Away Goals  850 non-null    float64
 13  Referee               850 non-null    object 
 14  Assistant 1           850 non-null    object 
 15  Assistant 2           8

## 3. Cargar datos

Al igual que hemos cargado los datos de los partidos en `df_partidos`, crear las variables `df_jugadores` y `df_copas` que contengan los dataframes correspondientes a la lectura de los csv `WorldCupPlayers` y `WorldCups` (usar `pd.read_csv)`.


In [ ]:
# Copiar solución de la actividad semanal correspondiente
df_jugadores = pd.read_csv("WorldCupPlayers.csv")
df_copas = pd.read_csv("WorldCups.csv")

## 4. Rango temporal

El dataframe `df_copas` contiene datos de todos los mundiales disputados. ¿Cuál es el año del mundial más antiguo disputado? ¿Y el año del mundial más reciente?

**Pista:** recuerda que podemos usar funciones de agregación (`count` para contar el número de casos, `mean` para calcular el valor medio, etc.) directamente sobre columnas o Series de pandas. Por ejemplo, el siguiente código nos muestra el mayor número de goles marcado en un mundial.

```python
df_copas["GoalsScored"].max()
```

**Opcional:** Una vez localizados los años de interés (más antiguo y más reciente) visualizar las filas completas correspondientes a cada año usando la función `.loc`.

In [ ]:
# Copiar solución de la actividad semanal correspondiente
print(f"El mundial más antiguo se jugó en el año {df_copas.Year.min()}, y el más reciente en el año {df_copas.Year.max()}")

El mundial más antiguo se jugó en el año 1930, y el más reciente en el año 2014


In [ ]:
df_copas.loc[df_copas.Year == df_copas.Year.min()]

Year  Country   Winner Runners-Up Third      Fourth  GoalsScored  \
0  1930  Uruguay  Uruguay  Argentina   USA  Yugoslavia           70   

   QualifiedTeams  MatchesPlayed Attendance  
0              13             18    590.549

In [ ]:
df_copas.loc[df_copas.Year == df_copas.Year.max()]

Year Country   Winner Runners-Up        Third  Fourth  GoalsScored  \
19  2014  Brazil  Germany  Argentina  Netherlands  Brazil          171   

    QualifiedTeams  MatchesPlayed Attendance  
19              32             64  3.386.810

---
# Ejercicios a realizar

Los ejercicios a continuación amplian el reporte comenzado en una actividad semanal anterior. Algunos de ellos son más complejos y son opcionales, pero se recomienda que al menos se intente resolverlos pues puntuará positivamente.

---

## 5. Asistencia de público

La columna `Attendance` indica la asistencia de público en cada mundial. Pandas ha cargado esta columna como tipo string, ya que los valores indicados contienen puntos como separador de millares. En primer lugar vamos a transformar esta columna para que no contenga el caracter `.`. Posteriormente transformaremos esta columna a formato numérico usando la función [pd.to_numeric](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_numeric.html).

**Nota**: la función `pd.read_csv` también permite hacer este proceso de una forma mucho más transparente, usando el argumento `thousands`. Ver la documentación [aquí](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html).

In [ ]:
df_copas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Year            20 non-null     int64 
 1   Country         20 non-null     object
 2   Winner          20 non-null     object
 3   Runners-Up      20 non-null     object
 4   Third           20 non-null     object
 5   Fourth          20 non-null     object
 6   GoalsScored     20 non-null     int64 
 7   QualifiedTeams  20 non-null     int64 
 8   MatchesPlayed   20 non-null     int64 
 9   Attendance      20 non-null     object
dtypes: int64(4), object(6)
memory usage: 1.7+ KB


In [ ]:
df_copas["Attendance"] = pd.to_numeric(df_copas["Attendance"].str.replace(".", ""))
# df_copas = pd.read_csv("WorldCups.csv", thousands=".")

<ipython-input-19-7b9d992eac5a>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df_copas["Attendance"] = pd.to_numeric(df_copas["Attendance"].str.replace(".", ""))


Teniendo en cuenta todos los datos de `df_copas`, ¿Cuál fué la mayor asistencia de un mundial? ¿Cuál es la asistencia media?

In [ ]:
from numpy import int64
# Completar
print(f"La mayor asistencia en un mundial {df_copas.Attendance.max()}, y la asistencia media {df_copas.Attendance.mean()}")

La mayor asistencia en un mundial 3587538, y la asistencia media 1872882.35


## 6. Asistencia de público en función del país

Sabemos que, dependiendo del país donde se celebre la copa del mundo, ésta tiene un mayor o menor impacto o relevancia. Vamos a calcular la media de asistencia agrupando por país. Finalmente, ordenar los resultados por orden decreciente de este número medio de asistentes.

**Pista 1:** Como se muestra en el siguiente ejemplo, recuerda que podemos combinar la función [groupby](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) con otras funciones de agregación como `mean`.

**Pista 2:** Recuerda que podemos usar la función [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html) para ordenar un DataFrame por una de sus columnas. Por ejemplo, el siguiente código muestra los mundiales ordenados de forma decreciente por el número de goles marcados:

```python
df_copas.sort_values("GoalsScored", ascending=False)
```

In [ ]:
df_personas = pd.DataFrame(
    [
        ("Juan", "male", 56),
        ("Laura", "female", 23),
        ("José", "male", 46),
        ("Rosa", "female", 26)
    ], columns=["name", "gender", "age"])

# edad media de todas las personas
edad_media = df_personas["age"].mean()
# edad media agrupando por género
edad_media_por_genero = df_personas.groupby("gender").mean()[["age"]]

print(f"La edad media de las {len(df_personas)} es de {edad_media} años.")
print(f"La edad media agrupada por género es:")
edad_media_por_genero

La edad media de las 4 es de 37.75 años.
La edad media agrupada por género es:


age
gender      
female  24.5
male    51.0

In [ ]:
# Completar
# media de asistencia agrupando por país
asistencia_media_por_pais = df_copas.groupby("Country").mean()[["Attendance"]]
print(f"Asistencia media por país:")
# ordenamos los valores de asistencia por orden decreciente
# se debe poner secuencial para que lo ordene sin volver a llamar a la variable
asistencia_media_por_pais.sort_values("Attendance", ascending=False)


Asistencia media por país:


Attendance
Country                 
USA            3587538.0
South Africa   3178856.0
Korea/Japan    2705197.0
Germany        2612596.0
Brazil         2216028.0
Spain          2109723.0
Mexico         1999003.0
France         1580400.0
England        1563135.0
Argentina      1545791.0
Italy          1439607.5
Chile           893172.0
Sweden          819810.0
Switzerland     768607.0
Uruguay         590549.0

In [ ]:
# Si ahora imprimimos la variable, no sale ordenada por número de asistencia, sino alfabético por país
print(f"Asistencia media por país:")
asistencia_media_por_pais


Asistencia media por país:


Attendance
Country                 
Argentina      1545791.0
Brazil         2216028.0
Chile           893172.0
England        1563135.0
France         1580400.0
Germany        2612596.0
Italy          1439607.5
Korea/Japan    2705197.0
Mexico         1999003.0
South Africa   3178856.0
Spain          2109723.0
Sweden          819810.0
Switzerland     768607.0
USA            3587538.0
Uruguay         590549.0

## 7. País con más victorias

Vamos a analizar el número de veces que un país ha ganado la copa del mundo. Transformar el DataFrame `df_copas` para que muestre los diferentes países que han ganado al menos una copa del mundo, y ordena el DataFrame por dicho número de forma descendente usando la función [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html).

¿Ves algo extraño o digno de comentar? Pista: analiza Alemania. **Hazlo en una nueva celda de tipo markdown.**

In [ ]:
df_copas

Year       Country      Winner      Runners-Up        Third  \
0   1930       Uruguay     Uruguay       Argentina          USA   
1   1934         Italy       Italy  Czechoslovakia      Germany   
2   1938        France       Italy         Hungary       Brazil   
3   1950        Brazil     Uruguay          Brazil       Sweden   
4   1954   Switzerland  Germany FR         Hungary      Austria   
5   1958        Sweden      Brazil          Sweden       France   
6   1962         Chile      Brazil  Czechoslovakia        Chile   
7   1966       England     England      Germany FR     Portugal   
8   1970        Mexico      Brazil           Italy   Germany FR   
9   1974       Germany  Germany FR     Netherlands       Poland   
10  1978     Argentina   Argentina     Netherlands       Brazil   
11  1982         Spain       Italy      Germany FR       Poland   
12  1986        Mexico   Argentina      Germany FR       France   
13  1990         Italy  Germany FR       Argentina        Italy   
14  1994           USA      Brazil           Italy       Sweden   
15  1998        France      France          Brazil      Croatia   
16  2002   Korea/Japan      Brazil         Germany       Turkey   
17  2006       Germany       Italy          France      Germany   
18  2010  South Africa       Spain     Netherlands      Germany   
19  2014        Brazil     Germany       Argentina  Netherlands   

            Fourth  GoalsScored  QualifiedTeams  MatchesPlayed  Attendance  
0       Yugoslavia           70              13             18      590549  
1          Austria           70              16             17      363000  
2           Sweden           84              15             18      375700  
3            Spain           88              13             22     1045246  
4          Uruguay          140              16             26      768607  
5       Germany FR          126              16             35      819810  
6       Yugoslavia           89              16             32      893172  
7     Soviet Union           89              16             32     1563135  
8          Uruguay           95              16             32     1603975  
9           Brazil           97              16             38     1865753  
10           Italy          102              16             38     1545791  
11          France          146              24             52     2109723  
12         Belgium          132              24             52     2394031  
13         England          115              24             52     2516215  
14        Bulgaria          141              24             52     3587538  
15     Netherlands          171              32             64     2785100  
16  Korea Republic          161              32             64     2705197  
17        Portugal          147              32             64     3359439  
18         Uruguay          145              32             64     3178856  
19          Brazil          171              32             64     3386810

In [ ]:
# Completar
# Se debe agrupar por winner y contar, ya que si lo agrupo por Country y hago conteo, me sale el número de veces que se ha celebreado
# la copa del mundo en el país.

numero_de_veces_pais_ganador = df_copas.groupby("Winner").count()["Year"]
print(f"Número de veces que un país ha sido ganador:")

# ordenamos los valores  por orden decreciente
numero_de_veces_pais_ganador.sort_values(ascending=False)

Número de veces que un país ha sido ganador:


Winner
Brazil        5
Italy         4
Germany FR    3
Argentina     2
Uruguay       2
England       1
France        1
Germany       1
Spain         1
Name: Year, dtype: int64

Alemania sale en el dataframe como Germany FR y como Germany. Esto es así porque en el 3 de octubre del año 1990 Alemania pasó de ser República Federal Alemana a formar Alemania unificada con la Alemania Oriental.
Por tanto, cuando ganó la copa en el año 2014, Alemania ya estaba unificada y por tanto ya no cuenta como Germany FR. Comentar que a pesar que en el año 1990 también ganó la copa Alemania, lo hizo como Germany FR porque la copa del mundo se celebró antes del 3 de octubre. 

---
# Parte opcional

Los ejercicios anteriores son obligatorios y puntuan hasta un máximo de 80 puntos sobre 100.

Los siguientes ejercicios trabajan con los dataframes cargados para extraer información concreta a partir de los datos. 
Estos ejercicios son opcionales, pero es recomendable intentar resolverlos para adquirir habilidades.

---

## Creación de un nuevo DataFrame

A continuación vamos a trabajar con el DataFrame `df_partidos`. Vamos a centrarnos en la información a nivel de selección, siendo indiferente si el equipo juega en casa o fuera. En este sentido, la información disponible no es la mejor para trabajar cómodamente. Por ello, es habitual realizar transformaciones a los datos en función de las preguntas que le queramos hacer. A continuación crearemos un nuevo DataFrame llamado `df_paises`que contendrá todos los paises involucrados en la copa del mundo y el año en el que participaron. Es decir, un país que haya jugado varias copas del mundo aparecerá varias veces:


||Year|Country|
|---|---|---|
|0|1930|France|
|1|1930|USA|
|...|...|..|
|37|1938|France|
|...|...|..|

**Nota:** este problema se puede afrontar como mínimo de dos maneras diferentes

1. Generando dos dataframes, uno donde la columna `Country` se corresponde con `Home Team Name`, y otro donde la columna `Country` se corresponde con `Away Team Name`. Finalmente, combinar ambos DataFrames en uno solo.
2. Usar la función [melt](https://pandas.pydata.org/docs/reference/api/pandas.melt.html) que nos permite convertir N columnas en un par de columnas llamadas `variable` y `value`, donde la primera es el nombre de una de las columnas y `value` el valor que tomó dicha variable.


In [ ]:
df_partidos

Year              Datetime                     Stage  \
0    1930.0  13 Jul 1930 - 15:00                    Group 1   
1    1930.0  13 Jul 1930 - 15:00                    Group 4   
2    1930.0  14 Jul 1930 - 12:45                    Group 2   
3    1930.0  14 Jul 1930 - 14:50                    Group 3   
4    1930.0  15 Jul 1930 - 16:00                    Group 1   
..      ...                   ...                       ...   
847  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
848  2014.0  08 Jul 2014 - 17:00                Semi-finals   
849  2014.0  09 Jul 2014 - 17:00                Semi-finals   
850  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
851  2014.0  13 Jul 2014 - 16:00                      Final   

                 Stadium             City Home Team Name  Home Team Goals  \
0                Pocitos      Montevideo          France              4.0   
1         Parque Central      Montevideo             USA              3.0   
2         Parque Central      Montevideo      Yugoslavia              2.0   
3                Pocitos      Montevideo         Romania              3.0   
4         Parque Central      Montevideo       Argentina              1.0   
..                   ...              ...            ...              ...   
847     Arena Fonte Nova        Salvador     Netherlands              0.0   
848     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
849   Arena de Sao Paulo       Sao Paulo     Netherlands              0.0   
850     Estadio Nacional        Brasilia          Brazil              0.0   
851  Estadio do Maracana  Rio De Janeiro         Germany              1.0   

     Away Team Goals Away Team Name                         Win conditions  \
0                1.0         Mexico                                          
1                0.0        Belgium                                          
2                1.0         Brazil                                          
3                1.0           Peru                                          
4                0.0         France                                          
..               ...            ...                                    ...   
847              0.0     Costa Rica  Netherlands win on penalties (4 - 3)    
848              7.0        Germany                                          
849              0.0      Argentina    Argentina win on penalties (2 - 4)    
850              3.0    Netherlands                                          
851              0.0      Argentina          Germany win after extra time    

     Attendance  Half-time Home Goals  Half-time Away Goals  \
0        4444.0                   3.0                   0.0   
1       18346.0                   2.0                   0.0   
2       24059.0                   2.0                   0.0   
3        2549.0                   1.0                   0.0   
4       23409.0                   0.0                   0.0   
..          ...                   ...                   ...   
847     51179.0                   0.0                   0.0   
848     58141.0                   0.0                   5.0   
849     63267.0                   0.0                   0.0   
850     68034.0                   0.0                   2.0   
851     74738.0                   0.0                   0.0   

                    Referee                  Assistant 1  \
0    LOMBARDI Domingo (URU)        CRISTOPHE Henry (BEL)   
1         MACIAS Jose (ARG)     MATEUCCI Francisco (URU)   
2       TEJADA Anibal (URU)      VALLARINO Ricardo (URU)   
3     WARNKEN Alberto (CHI)          LANGENUS Jean (BEL)   
4       REGO Gilberto (BRA)         SAUCEDO Ulises (BOL)   
..                      ...                          ...   
847   Ravshan IRMATOV (UZB)  RASULOV Abduxamidullo (UZB)   
848   RODRIGUEZ Marco (MEX)      TORRENTERA Marvin (MEX)   
849      C�neyt �AKIR (TUR)         DURAN Bahattin (TUR)   
850   HAIMOUDI Djamel (ALG)    

In [ ]:
#alternativa 1
df_paises = (
    pd.concat([
        df_partidos[["Year", "Home Team Name"]].rename(columns={"Home Team Name": "Country"}),
        df_partidos[["Year", "Away Team Name"]].rename(columns={"Away Team Name": "Country"})
    ])
)
#alternativa 2
#df_paises = (
#    df_partidos
#    .melt(id_vars=["Year"], value_vars=["Home Team Name", "Away Team Name"])
#    .rename(columns={"value": "Country"})
#    [["Year", "Country"]]
#)

In [ ]:
df_paises.head()

Year     Country
0  1930.0      France
1  1930.0         USA
2  1930.0  Yugoslavia
3  1930.0     Romania
4  1930.0   Argentina

## 8. Número de países

Vamos a descubrir cuántos países han participado, al menos una vez, en una copa del mundo. Para ello vamos a trabajar con el DataFrame construido anteriormente y a usar la función de agregación [nunique](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.nunique.html).

In [ ]:
# Completar
# Count number of distinct elements in specified axis.
df_paises.nunique(axis=0, dropna=True)

Year       20
Country    83
dtype: int64

## 9. Número de partidos por país

Ahora vamos a descubrir cuántos partidos ha jugado cada país. Operar con el DataFrame para que muestre un país por fila y una columna que indique el número de partidos que ha jugado. Ordenar el DataFrame por dicha columna de forma descendente.

In [ ]:
# Completar
numero_de_partidos_por_pais = df_paises.groupby("Country").count()["Year"]
# ordenamos los valores  por orden decreciente
numero_de_partidos_por_pais.sort_values(ascending=False)


Country
Brazil                      108
Italy                        83
Argentina                    81
Germany FR                   62
England                      62
                           ... 
Iraq                          3
Kuwait                        3
Serbia                        3
rn">United Arab Emirates      3
Dutch East Indies             1
Name: Year, Length: 83, dtype: int64

In [ ]:
numero_de_partidos_por_pais = df_paises.groupby("Country").count()
numero_de_partidos_por_pais

Year
Country                         
Algeria                       12
Angola                         3
Argentina                     81
Australia                     13
Austria                       29
...                          ...
rn">Bosnia and Herzegovina     3
rn">Republic of Ireland       13
rn">Serbia and Montenegro      3
rn">Trinidad and Tobago        3
rn">United Arab Emirates       3

[83 rows x 1 columns]

## 10. Partidos jugados por España

Vamos a trabajar a continuación con el DataFrame `df_partidos`. Seleccionar aquellos partidos (filas del dataframe) en los que España ha jugado como `Home Team`, es decir, donde la columna `Home Team` tiene el valor `España`. Recuerda que para la operación de filtrado hay que usar la función [.loc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html).


A continuación, haz una nueva selección en los que España ha jugado como `Away Team`, es decir, donde la columna `Away Team` tiene el valor `España`.

**Opcional:** ¿Podrías hacer una sola selección donde España haya jugado como `Home Team` o como `Away Team`?. Pista: consulta la documentación del operador de [disyunción](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#boolean-indexing).

In [ ]:
df_partidos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 850 entries, 0 to 851
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  850 non-null    float64
 1   Datetime              850 non-null    object 
 2   Stage                 850 non-null    object 
 3   Stadium               850 non-null    object 
 4   City                  850 non-null    object 
 5   Home Team Name        850 non-null    object 
 6   Home Team Goals       850 non-null    float64
 7   Away Team Goals       850 non-null    float64
 8   Away Team Name        850 non-null    object 
 9   Win conditions        850 non-null    object 
 10  Attendance            850 non-null    float64
 11  Half-time Home Goals  850 non-null    float64
 12  Half-time Away Goals  850 non-null    float64
 13  Referee               850 non-null    object 
 14  Assistant 1           850 non-null    object 
 15  Assistant 2           8

In [ ]:
# Completar
# utilizamos la función loc para filtrar por el campo España dentro de la columna HomeTeam Name
print ("Partidos que España ha jugado en casa")
df_partidos.loc[df_partidos.loc[:,'Home Team Name'] == 'Spain']

Partidos que España ha jugado en casa


Year               Datetime              Stage  \
23   1934.0   27 May 1934 - 16:30   Preliminary round   
55   1950.0   25 Jun 1950 - 15:00             Group 2   
60   1950.0   29 Jun 1950 - 15:00             Group 2   
64   1950.0   02 Jul 1950 - 15:00             Group 2   
149  1962.0   03 Jun 1962 - 15:00             Group 3   
177  1966.0   15 Jul 1966 - 19:30             Group 2   
291  1978.0   11 Jun 1978 - 13:45             Group 3   
316  1982.0   16 Jun 1982 - 21:00             Group 5   
328  1982.0   20 Jun 1982 - 21:00             Group 5   
355  1982.0   05 Jul 1982 - 21:00             Group 2   
362  1986.0   01 Jun 1986 - 12:00             Group D   
406  1986.0   22 Jun 1986 - 16:00      Quarter-finals   
454  1990.0   26 Jun 1990 - 17:00         Round of 16   
464  1994.0   17 Jun 1994 - 19:30             Group C   
501  1994.0   02 Jul 1994 - 16:30         Round of 16   
523  1998.0   13 Jun 1998 - 14:30             Group D   
539  1998.0   19 Jun 1998 - 21:00             Group D   
554  1998.0   24 Jun 1998 - 21:00             Group D   
587  2002.0   02 Jun 2002 - 20:30             Group B   
602  2002.0  07 June 2002 - 18:00             Group B   
630  2002.0   16 Jun 2002 - 20:30         Round of 16   
639  2002.0   22 Jun 2002 - 15:30      Quarter-finals   
658  2006.0   14 Jun 2006 - 15:00             Group H   
675  2006.0   19 Jun 2006 - 21:00             Group H   
699  2006.0   27 Jun 2006 - 21:00         Round of 16   
723  2010.0   16 Jun 2010 - 16:00             Group H   
739  2010.0   21 Jun 2010 - 20:30             Group H   
763  2010.0   29 Jun 2010 - 20:30         Round of 16   
774  2014.0   13 Jun 2014 - 16:00             Group B   
790  2014.0   18 Jun 2014 - 16:00             Group B   

                                       Stadium             City  \
23                              Luigi Ferraris           Genoa    
55                            Durival de Brito        Curitiba    
60   Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
64   Maracan� - Est�dio Jornalista M�rio Filho  Rio De Janeiro    
149                          Estadio Sausalito    Vina Del Mar    
177                               Hillsborough       Sheffield    
291                            Jose Amalfitani    Buenos Aires    
316                              Luis Casanova        Valencia    
328                              Luis Casanova        Valencia    
355                          Santiago Bernabeu          Madrid    
362                                    Jalisco     Guadalajara    
406                                 Cuauhtemoc          Puebla    
454                     Marc Antonio Bentegodi          Verona    
464                                Cotton Bowl          Dallas    
501                                RFK Stadium   Washington Dc    
523                               La Beaujoire          Nantes    
539                    Stade Geoffroy Guichard   Saint-Etienne    
554                       Stade Felix Bollaert            Lens    
587                  Gwangju World Cup Stadium         Gwangju    
602                   Jeonju World Cup Stadium          Jeonju    
630                    Suwon World Cup Stadium           Suwon    
639                  Gwangju World Cup Stadium         Gwangju    
658                             Zentralstadion         Leipzig    
675                   Gottlieb-Daimler-Stadion       Stuttgart    
699            FIFA World Cup Stadium, Hanover         Hanover    
723                             Durban Stadium          Durban    
739                         Ellis Park Stadium    Johannesburg    
763                          Cape Town Stadium       Cape Town    
774                           Arena Fonte Nova        Salvador    
790                        Estadio do Maracana  Rio De Janeiro    

    Home Team Name  Home Team Goals  Away Team Goals           Away Team Name  \
23           Spain              3.0              1.0                   Brazil   

In [ ]:
# utilizamos la función loc para filtrar por el campo España dentro de la columna AwayTeam Name
print ("Partidos que España ha jugado fuera de casa")
df_partidos.loc[df_partidos.loc[:,'Away Team Name'] == 'Spain']

Partidos que España ha jugado fuera de casa


Year              Datetime           Stage  \
28   1934.0  31 May 1934 - 16:30   Quarter-finals   
30   1934.0  01 Jun 1934 - 16:30   Quarter-finals   
69   1950.0  09 Jul 1950 - 15:00          Group 6   
71   1950.0  13 Jul 1950 - 15:00          Group 6   
73   1950.0  16 Jul 1950 - 15:00          Group 6   
141  1962.0  31 May 1962 - 15:00          Group 3   
153  1962.0  06 Jun 1962 - 15:00          Group 3   
174  1966.0  13 Jul 1966 - 19:30          Group 2   
190  1966.0  20 Jul 1966 - 19:30          Group 2   
275  1978.0  03 Jun 1978 - 13:45          Group 3   
282  1978.0  07 Jun 1978 - 13:45          Group 3   
343  1982.0  25 Jun 1982 - 21:00          Group 5   
351  1982.0  02 Jul 1982 - 21:00          Group 2   
378  1986.0  07 Jun 1986 - 12:00          Group D   
393  1986.0  12 Jun 1986 - 12:00          Group D   
403  1986.0  18 Jun 1986 - 16:00      Round of 16   
423  1990.0  13 Jun 1990 - 17:00          Group E   
434  1990.0  17 Jun 1990 - 21:00          Group E   
445  1990.0  21 Jun 1990 - 17:00          Group E   
476  1994.0  21 Jun 1994 - 16:00          Group C   
491  1994.0  27 Jun 1994 - 16:00          Group C   
508  1994.0  09 Jul 1994 - 12:00   Quarter-finals   
619  2002.0  12 Jun 2002 - 20:30          Group B   
689  2006.0  23 Jun 2006 - 16:00          Group H   
754  2010.0  25 Jun 2010 - 20:30          Group H   
767  2010.0  03 Jul 2010 - 20:30   Quarter-finals   
769  2010.0  07 Jul 2010 - 20:30      Semi-finals   
771  2010.0  11 Jul 2010 - 20:30            Final   
804  2014.0  23 Jun 2014 - 13:00          Group B   

                                       Stadium               City  \
28                              Giovanni Berta          Florence    
30                              Giovanni Berta          Florence    
69                                    Pacaembu         Sao Paulo    
71   Maracan� - Est�dio Jornalista M�rio Filho    Rio De Janeiro    
73                                    Pacaembu         Sao Paulo    
141                          Estadio Sausalito      Vina Del Mar    
153                          Estadio Sausalito      Vina Del Mar    
174                                 Villa Park        Birmingham    
190                                 Villa Park        Birmingham    
275                            Jose Amalfitani      Buenos Aires    
282                 Estadio Jos� Mar�a Minella     Mar Del Plata    
343                              Luis Casanova          Valencia    
351                          Santiago Bernabeu            Madrid    
378                              Tres de Marzo       Guadalajara    
393                                Tecnologico         Monterrey    
403                        Estadio Corregidora         Queretaro    
423                                Dacia Arena             Udine    
434                                Dacia Arena             Udine    
445                     Marc Antonio Bentegodi            Verona    
476                              Soldier Field           Chicago    
491                              Soldier Field           Chicago    
508                            Foxboro Stadium            Boston    
619                  Daejeon World Cup Stadium           Daejeon    
689                       Fritz-Walter-Stadion    Kaiserslautern    
754                    Loftus Versfeld Stadium  Tshwane/Pretoria    
767                         Ellis Park Stadium      Johannesburg    
769                             Durban Stadium            Durban    
771                        Soccer City Stadium      Johannesburg    
804                           Arena da Baixada          Curitiba    

       Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
28              Italy              1.0              1.0          Spain   
30              Italy              1.0              0.0          Spain   
69            Uruguay              2.0              2.0          Spain   
71             Brazil              6.0           

In [ ]:
# Opcional: utilizamos el operador booleano | que significa "or"

(
    df_partidos
    .loc[lambda df: df["Home Team Name"] == "Spain"]
).head(2)

Year              Datetime              Stage           Stadium  \
23  1934.0  27 May 1934 - 16:30   Preliminary round    Luigi Ferraris   
55  1950.0  25 Jun 1950 - 15:00             Group 2  Durival de Brito   

         City Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
23     Genoa           Spain              3.0              1.0         Brazil   
55  Curitiba           Spain              3.0              1.0            USA   

   Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
23                    21000.0                   3.0                   0.0   
55                     9511.0                   0.0                   1.0   

                Referee                 Assistant 1  \
23  BIRLEM Alfred (GER)      CARMINATI Ettore (ITA)   
55    VIANA Mario (BRA)  DA COSTA VIEIRA Jose (POR)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
23     IVANCSICS Mihaly (HUN)    204.0   1111.0                ESP   
55  DE LA SALLE Charles (FRA)    208.0   1208.0                ESP   

   Away Team Initials  
23                BRA  
55                USA

In [ ]:
(
    df_partidos
    .loc[lambda df: df["Away Team Name"] == "Spain"]
).head(2)

Year              Datetime           Stage         Stadium       City  \
28  1934.0  31 May 1934 - 16:30   Quarter-finals  Giovanni Berta  Florence    
30  1934.0  01 Jun 1934 - 16:30   Quarter-finals  Giovanni Berta  Florence    

   Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
28          Italy              1.0              1.0          Spain   
30          Italy              1.0              0.0          Spain   

   Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
28                    35000.0                   0.0                   0.0   
30                    43000.0                   1.0                   0.0   

              Referee             Assistant 1             Assistant 2  \
28  BAERT Louis (BEL)   ZENISEK Bohumil (TCH)  IVANCSICS Mihaly (HUN)   
30  MERCET Rene (SUI)  IVANCSICS Mihaly (HUN)   ZENISEK Bohumil (TCH)   

    RoundID  MatchID Home Team Initials Away Team Initials  
28    418.0   1122.0                ITA                ESP  
30    418.0   1123.0                ITA                ESP

In [ ]:
(
    df_partidos
    .loc[lambda df: (df["Home Team Name"] == "Spain") | (df["Away Team Name"] == "Spain")]
).head(2)

Year              Datetime              Stage         Stadium  \
23  1934.0  27 May 1934 - 16:30   Preliminary round  Luigi Ferraris   
28  1934.0  31 May 1934 - 16:30      Quarter-finals  Giovanni Berta   

         City Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
23     Genoa           Spain              3.0              1.0         Brazil   
28  Florence           Italy              1.0              1.0          Spain   

   Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
23                    21000.0                   3.0                   0.0   
28                    35000.0                   0.0                   0.0   

                Referee             Assistant 1             Assistant 2  \
23  BIRLEM Alfred (GER)  CARMINATI Ettore (ITA)  IVANCSICS Mihaly (HUN)   
28    BAERT Louis (BEL)   ZENISEK Bohumil (TCH)  IVANCSICS Mihaly (HUN)   

    RoundID  MatchID Home Team Initials Away Team Initials  
23    204.0   1111.0                ESP                BRA  
28    418.0   1122.0                ITA                ESP

## Calcular el vencedor de cada partido

El DataFrame `df_partidos` tiene la información de cada partido, incluyendo los goles marcados por cada equipo, pero no disponemos de una columna que indique si un equipo ganó o perdió. Vamos a crear una columna llamada `Ganador` que indique el nombre del equipo vencedor (con más goles). También podemos encontrar empates, en cuyo caso usaremos `pd.NA` para indicar que desconocemos el vencedor de dicho partido.

**Nota:** Se puede resolver realizando varias asignaciones, o una sola utilizando la función [pd.where](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.where.html).

In [ ]:
# alternativa 1
df_partidos["Ganador"] = df_partidos["Home Team Name"]  # ponemos la variable del pais ganador que es el mismo que Home Team Goals, para inializar
df_partidos.loc[df_partidos["Home Team Goals"] < df_partidos["Away Team Goals"], "Ganador"] = df_partidos["Away Team Name"]
df_partidos.loc[df_partidos["Home Team Goals"] == df_partidos["Away Team Goals"], "Ganador"] = pd.NA

# alternativa 2
#df_partidos["Ganador"] = (
#    df_partidos["Home Team Name"].where(
#        df_partidos["Home Team Goals"] > df_partidos["Away Team Goals"],
#        df_partidos["Away Team Name"].where(df_partidos["Home Team Goals"] < df_partidos["Away Team Goals"], pd.NA)
#    )
#)

<ipython-input-38-0597d4970d51>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_partidos["Ganador"] = df_partidos["Home Team Name"]  # ponemos la variable del pais ganador que es el mismo que Home Team Goals, para inializar
/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)
/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be se

In [ ]:

df_partidos_vencedor= df_partidos[["Year", "Stage", "Home Team Name", "Away Team Name", "Ganador"]] # las columnas se ponen sin corchetes
df_partidos_vencedor

Year                     Stage Home Team Name Away Team Name  \
0    1930.0                   Group 1         France         Mexico   
1    1930.0                   Group 4            USA        Belgium   
2    1930.0                   Group 2     Yugoslavia         Brazil   
3    1930.0                   Group 3        Romania           Peru   
4    1930.0                   Group 1      Argentina         France   
..      ...                       ...            ...            ...   
847  2014.0            Quarter-finals    Netherlands     Costa Rica   
848  2014.0               Semi-finals         Brazil        Germany   
849  2014.0               Semi-finals    Netherlands      Argentina   
850  2014.0  Play-off for third place         Brazil    Netherlands   
851  2014.0                     Final        Germany      Argentina   

         Ganador  
0         France  
1            USA  
2     Yugoslavia  
3        Romania  
4      Argentina  
..           ...  
847         <NA>  
848      Germany  
849         <NA>  
850  Netherlands  
851      Germany  

[850 rows x 5 columns]

In [ ]:
(
    df_partidos
    .dropna(subset=["Ganador"])
    .groupby("Ganador")
    .count()
    .sort_values("Year", ascending=False)
)

Year  Datetime  Stage  Stadium  City  \
Ganador                                                            
Brazil                        71        71     71       71    71   
Italy                         45        45     45       45    45   
Argentina                     44        44     44       44    44   
Germany FR                    36        36     36       36    36   
Germany                       32        32     32       32    32   
...                          ...       ...    ...      ...   ...   
Cuba                           1         1      1        1     1   
Korea DPR                      1         1      1        1     1   
Wales                          1         1      1        1     1   
rn">Bosnia and Herzegovina     1         1      1        1     1   
Iran                           1         1      1        1     1   

                            Home Team Name  Home Team Goals  Away Team Goals  \
Ganador                                                                        
Brazil                                  71               71               71   
Italy                                   45               45               45   
Argentina                               44               44               44   
Germany FR                              36               36               36   
Germany                                 32               32               32   
...                                    ...              ...              ...   
Cuba                                     1                1                1   
Korea DPR                                1                1                1   
Wales                                    1                1                1   
rn">Bosnia and Herzegovina               1                1                1   
Iran                                     1                1                1   

                            Away Team Name  Win conditions  Attendance  \
Ganador                                                                  
Brazil                                  71              71          71   
Italy                                   45              45          45   
Argentina                               44              44          44   
Germany FR                              36              36          36   
Germany                                 32              32          32   
...                                    ...             ...         ...   
Cuba                                     1               1           1   
Korea DPR                                1               1           1   
Wales                                    1               1           1   
rn">Bosnia and Herzegovina               1               1           1   
Iran                                     1               1           1   

                            Half-time Home Goals  Half-time Away Goals  \
Ganador                                                                  
Brazil                                        71                    71   
Italy                                         45                    45   
Argentina                                     44                    44   
Germany FR                                    36                    36   
Germany                                       32                    32   
...                                          ...                   ...   
Cuba                                           1                     1   
Korea DPR                                      1                     1   
Wales                                          1                     1   
rn">Bosnia and Herzegovina                     1                     1   
Iran                                           1                     1   

                            Referee  Assistant 1  Assistant 2  RoundID  \
Ganador                                                                  
Brazil                      

## 11. Número de victorias por país

Calcular para cada país el número de partidos ganados. Para ello, usar la función [dropna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) para eliminar las filas correspondientes a partidos donde la columna `Ganador` es un valor perdido (`na`). Ordenar los países por número de victorias de forma descendente.

In [ ]:
# Completar
# Eliminamos las filas con algún valor perdido
df_partidos = df_partidos.dropna()
df_partidos ["Ganador"]

0           France
1              USA
2       Yugoslavia
3          Romania
4        Argentina
          ...     
845         Brazil
846      Argentina
848        Germany
850    Netherlands
851        Germany
Name: Ganador, Length: 660, dtype: object

In [ ]:
df_partidos.loc[df_partidos.isna().any(axis=1)].head()

Empty DataFrame
Columns: [Year, Datetime, Stage, Stadium, City, Home Team Name, Home Team Goals, Away Team Goals, Away Team Name, Win conditions, Attendance, Half-time Home Goals, Half-time Away Goals, Referee, Assistant 1, Assistant 2, RoundID, MatchID, Home Team Initials, Away Team Initials, Ganador]
Index: []

[0 rows x 21 columns]

In [ ]:
# Se debe agrupar por Ganador y contar. También debo incluir la columna "Year" para que me salga una columna sólo

numero_de_veces_pais_ganador = df_partidos.groupby("Ganador").count()["Year"]
print(f"Número de veces que un país ha ganado un partido:")
# ordenamos los valores  por orden decreciente

numero_de_veces_pais_ganador.sort_values(ascending=False)


Número de veces que un país ha ganado un partido:


Ganador
Brazil                        71
Italy                         45
Argentina                     44
Germany FR                    36
Germany                       32
                              ..
Cuba                           1
Korea DPR                      1
Wales                          1
rn">Bosnia and Herzegovina     1
Iran                           1
Name: Year, Length: 64, dtype: int64

In [ ]:
# Resultado de enfrentanmientos
(
    df_partidos
    .dropna(subset=["Ganador"])
    .groupby("Ganador")
    .count()
    .sort_values("Year", ascending=False)
)

Year  Datetime  Stage  Stadium  City  \
Ganador                                                            
Brazil                        71        71     71       71    71   
Italy                         45        45     45       45    45   
Argentina                     44        44     44       44    44   
Germany FR                    36        36     36       36    36   
Germany                       32        32     32       32    32   
...                          ...       ...    ...      ...   ...   
Cuba                           1         1      1        1     1   
Korea DPR                      1         1      1        1     1   
Wales                          1         1      1        1     1   
rn">Bosnia and Herzegovina     1         1      1        1     1   
Iran                           1         1      1        1     1   

                            Home Team Name  Home Team Goals  Away Team Goals  \
Ganador                                                                        
Brazil                                  71               71               71   
Italy                                   45               45               45   
Argentina                               44               44               44   
Germany FR                              36               36               36   
Germany                                 32               32               32   
...                                    ...              ...              ...   
Cuba                                     1                1                1   
Korea DPR                                1                1                1   
Wales                                    1                1                1   
rn">Bosnia and Herzegovina               1                1                1   
Iran                                     1                1                1   

                            Away Team Name  Win conditions  Attendance  \
Ganador                                                                  
Brazil                                  71              71          71   
Italy                                   45              45          45   
Argentina                               44              44          44   
Germany FR                              36              36          36   
Germany                                 32              32          32   
...                                    ...             ...         ...   
Cuba                                     1               1           1   
Korea DPR                                1               1           1   
Wales                                    1               1           1   
rn">Bosnia and Herzegovina               1               1           1   
Iran                                     1               1           1   

                            Half-time Home Goals  Half-time Away Goals  \
Ganador                                                                  
Brazil                                        71                    71   
Italy                                         45                    45   
Argentina                                     44                    44   
Germany FR                                    36                    36   
Germany                                       32                    32   
...                                          ...                   ...   
Cuba                                           1                     1   
Korea DPR                                      1                     1   
Wales                                          1                     1   
rn">Bosnia and Herzegovina                     1                     1   
Iran                                           1                     1   

                            Referee  Assistant 1  Assistant 2  RoundID  \
Ganador                                                                  
Brazil                      

## Opcional 1

Calcular para cada país el ratio de victorias, calculado como el número de partidos ganados dividido por el número de partidos jugados. Ordenar los países por ratio de forma descendente.

**Pista:** Si obtenemos dos dataframes, uno con las victorias por cada país y otro con el número total de partidos jugados, podemos utilizar la función [pd.merge](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) para combinar ambos en un solo DataFrame, y despues crear una nueva columna que sea el ratio.

In [ ]:
# Completar
#DataFrame.merge(right, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, sort=False, suffixes=('_x', '_y'), copy=True, indicator=False, validate=None)[source]
#numero_de_partidos_por_pais.merge(numero_de_veces_pais_ganador, left_on= "num_partidos" , right_on= "ganador")

In [ ]:
# Ordenar por ratio
victorias = (
    df_partidos
    .dropna(subset=["Ganador"])
    .groupby("Ganador")
    .count()
    .rename(columns={"Year":"Victorias"})
    [["Victorias"]]
)

jugados = (
    df_paises
    .groupby("Country")
    .count()
    .rename(columns={"Year": "Jugados"})
    [["Jugados"]]
)
ratio_victorias = pd.merge(jugados, victorias, left_index=True, right_index=True, how="inner")
ratio_victorias["Ratio"] = ratio_victorias["Victorias"] / ratio_victorias["Jugados"]

ratio_victorias.sort_values("Ratio", ascending=False)

Jugados  Victorias     Ratio
Germany                       46         32  0.695652
Brazil                       108         71  0.657407
Germany FR                    62         36  0.580645
Argentina                     81         44  0.543210
Italy                         83         45  0.542169
...                          ...        ...       ...
Morocco                       13          2  0.153846
rn">Republic of Ireland       13          2  0.153846
Korea DPR                      7          1  0.142857
Bulgaria                      26          3  0.115385
Tunisia                       12          1  0.083333

[64 rows x 3 columns]

## Opcional 2

¿Se te ocurre alguna pregunta adicional que hacerle a los datos? Utiliza una celda de tipo markdown para documentarla y otra de código para implementarla y ver así la respuesta.

### ¿Cuál es la lista de los jugadores que más partidos hanjugado en el mundial?

In [ ]:
df_jugadores = pd.read_csv("WorldCupPlayers.csv")
df_jugadores.head()


RoundID  MatchID Team Initials           Coach Name Line-up  Shirt Number  \
0      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
1      201     1096           MEX     LUQUE Juan (MEX)       S             0   
2      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
3      201     1096           MEX     LUQUE Juan (MEX)       S             0   
4      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   

        Player Name Position Event  
0       Alex THEPOT       GK   NaN  
1   Oscar BONFIGLIO       GK   NaN  
2  Marcel LANGILLER      NaN  G40'  
3      Juan CARRENO      NaN  G70'  
4   Ernest LIBERATI      NaN   NaN

In [ ]:
# Se debe agrupar por Jugador y contar. También debo incluir la columna "MatchID" para que me salga una columna sólo

numero_de_partidos_jugador = df_jugadores.groupby("Player Name").count()
print(f"Número de partidos que ha jugado un jugador:")
# ordenamos los valores  por orden decreciente
numero_de_partidos_jugador
#numero_de_partidos_jugador.sort_values(ascending=False)



Número de partidos que ha jugado un jugador:


RoundID  MatchID  Team Initials  Coach Name  Line-up  \
Player Name                                                         
?URI?I?            3        3              3           3        3   
A BAUTISTA         4        4              4           4        4   
A COLE            14       14             14          14       14   
A GUARDADO         4        4              4           4        4   
A MEDINA           4        4              4           4        4   
...              ...      ...            ...         ...      ...   
�IGI?              3        3              3           3        3   
�STENSTAD          4        4              4           4        4   
�ULER              3        3              3           3        3   
�UNJI?             3        3              3           3        3   
�ZIL              18       18             18          18       18   

             Shirt Number  Position  Event  
Player Name                                 
?URI?I?                 3         3      0  
A BAUTISTA              4         0      1  
A COLE                 14         0      3  
A GUARDADO              4         0      3  
A MEDINA                4         0      0  
...                   ...       ...    ...  
�IGI?                   3         0      2  
�STENSTAD               4         0      1  
�ULER                   3         0      1  
�UNJI?                  3         0      0  
�ZIL                   18         0     13  

[7663 rows x 8 columns]

### ¿Cuáles son los jugadores españoles que han jugado en copa del mundo"

In [ ]:
jugadores_espanoles = df_jugadores.loc[df_jugadores.loc[:,'Team Initials'] == 'ESP']
print(f"Número de partidos que ha jugado un jugador español:")
jugadores_espanoles.groupby("Player Name").count()




Número de partidos que ha jugado un jugador español:


RoundID  MatchID  Team Initials  Coach Name  Line-up  \
Player Name                                                            
A. INIESTA            7        7              7           7        7   
A. LOPEZ              4        4              4           4        4   
A.INIESTA             7        7              7           7        7   
ABELARDO              8        8              8           8        8   
ADELARDO              6        6              6           6        6   
...                 ...      ...            ...         ...      ...   
Vicente ASENSI        6        6              6           6        6   
XAVI                 19       19             19          19       19   
ZABALO                3        3              3           3        3   
ZARRA                 6        6              6           6        6   
ZUBIZARRETA          12       12             12          12       12   

                Shirt Number  Position  Event  
Player Name                                    
A. INIESTA                 7         0      0  
A. LOPEZ                   4         0      0  
A.INIESTA                  7         0      3  
ABELARDO                   8         0      3  
ADELARDO                   6         0      1  
...                      ...       ...    ...  
Vicente ASENSI             6         0      0  
XAVI                      19         0      6  
ZABALO                     3         0      0  
ZARRA                      6         1      4  
ZUBIZARRETA               12        12      1  

[239 rows x 8 columns]

In [ ]:
jugadores_espanoles


RoundID    MatchID Team Initials                Coach Name Line-up  \
895        204       1111           ESP       GARCIA Amadeo (ESP)       S   
897        204       1111           ESP       GARCIA Amadeo (ESP)       S   
899        204       1111           ESP       GARCIA Amadeo (ESP)       S   
901        204       1111           ESP       GARCIA Amadeo (ESP)       S   
903        204       1111           ESP       GARCIA Amadeo (ESP)       S   
...        ...        ...           ...                       ...     ...   
35613   255931  300186467           ESP  DEL BOSQUE Vicente (ESP)       N   
35615   255931  300186467           ESP  DEL BOSQUE Vicente (ESP)       N   
35617   255931  300186467           ESP  DEL BOSQUE Vicente (ESP)       N   
35619   255931  300186467           ESP  DEL BOSQUE Vicente (ESP)       N   
35621   255931  300186467           ESP  DEL BOSQUE Vicente (ESP)       N   

       Shirt Number     Player Name Position      Event  
895               0  Ricardo ZAMORA      GKC        NaN  
897               0       CILAURREN      NaN        NaN  
899               0         CIRIACO      NaN        NaN  
901               0     I QUINCOCES      NaN        NaN  
903               0       GOROSTIZA      NaN        NaN  
...             ...             ...      ...        ...  
35613            13            MATA      NaN  I56' G82'  
35615            16          SERGIO      NaN        NaN  
35617            19     DIEGO COSTA      NaN        NaN  
35619            21           SILVA      NaN       I83'  
35621            22     AZPILICUETA      NaN        NaN  

[1317 rows x 9 columns]